Cleanup src/solver.rs

xfbs opened the pull request:
This PR does some little refactoring of the `resolve` method.

- I rearranged the code so that it reduces nesting. I do that by rearranging conditionals so that the error case returns early and then paves the way for the (non-indented) success path.
- I give booleans a name (I tried, hope the names make sense) so that if statements "make more sense"
- I turn an "expect" into an error and remove an `unwrap()` by using a match statement.

xfbs committed:
refactor: clean up src/solver.rs

- Reduces nesting, to increase legibility of code
- Turns expect into error
- Removes unwrap()

Eh2406 commented on src/solver.rs:
```diff
@@ -94,36 +95,38 @@ pub fn resolve<P: Package, VS: VersionSet>(
-                )
-            });
-        }
+        let potential_packages = match state.partial_solution.potential_packages() {
```
Looking at the code as you have it here, it seems like a perfect place for the new `let ... else` syntax.

xfbs replied:
Ha! Never used that one before. Pretty neat. I don't like how it makes the initial line quite long, but I do like how it removes one level of indentation from the body.

Eh2406 replied:
Talking about syntax is so confusing. No there is a newer and even more concise syntax https://rust-lang.github.io/rfcs/3137-let-else.html

Something like
```rust
        let Some(packages) = state.partial_solution.potential_packages() else {
            return state.partial_solution.extract_solution().ok_or_else(|| {
                PubGrubError::Failure(
                    "How did we end up with no package to choose but no solution?".into(),
                )
            });
        };
```

xfbs replied:
Sweet, I've just re-pushed it using that syntax :) That is definitely more concise.

Eh2406 reviewed (approved):
Thanks for the improvements!

mpizenberg reviewed (commented):
Thanks for the PR! At first read I like most of it though I have a few remarks. We tend to be as specific as possible and favor reader over writer, so I'm not a super fan of unqualified import (for the `log` module).
Another nitpick, I'd prefer avoiding shadowing of variable (`dependencies`).

xfbs commented:
> Thanks for the PR! 

Happy to help, awesome crate :) We use this in production btw!

> At first read I like most of it though I have a few remarks. We tend to be as specific as possible and favor reader over writer, so I'm not a super fan of unqualified import (for the log module).

I definitely agree with you! I tend to always explicitly include anything I use, with only two exceptions: `log::*` and `tracing::*` (never both, just either one of them ofc), just because these crates are so useful everywhere and it's nice to be able to just plop a `debug!()` somewhere. But this is a personal preference, so I changed to explicitly importing them (I hope that is what you prefer).

> Another nitpick, I'd prefer avoiding shadowing of variable (dependencies).

That is one feature of Rust I love :D but I get you, so I changed it, I hope my naming makes sense (I'm not too familiar with the codebase).

Feel free to also just take some part of my PR and throw out other parts! Style is always very subjective, I don't want to enforce my ideas, just sharing something that might be useful.

Cheers!

Eh2406 commented:
@mpizenberg are we good to merge now?

mpizenberg commented:
If it's good for you, let me final check Sunday and I'll merge. If you
don't here from me on Monday, consider it's good to merge :)

On Fri, Jun 9, 2023, 18:57 Jacob Finkelman ***@***.***> wrote:

> @mpizenberg <https://github.com/mpizenberg> are we good to merge now?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pubgrub-rs/pubgrub/pull/127#issuecomment-1584887945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWFOCKEYUGZOBJ6ISYM7J3XKNIYHANCNFSM6AAAAAAXWEQZYI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mpizenberg commented:
Feels good to get rid of another `unwrap` and `expect` :)
Thanks again for the improvements.
